contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,207 @@
     }
 
-    /**
-      * validate the request:
-      * if this is a constructor call, make sure it is a known account.
-      * verify the sender has enough tokens.
-      * (since the paymaster is also the token, there is no notion of "approval")
-      */
-    function _validatePaymasterUserOp(UserOperation calldata userOp, bytes32 /*userOpHash*/, uint256 requiredPreFund)
-    internal view override returns (bytes memory context, uint256 validationData) {
-        uint256 tokenPrefund = getTokenValueOfEth(requiredPreFund);
-
-        // verificationGasLimit is dual-purposed, as gas limit for postOp. make sure it is high enough
-        // make sure that verificationGasLimit is high enough to handle postOp
-        require(userOp.verificationGasLimit > COST_OF_POST, "TokenPaymaster: gas too low for postOp");
-
-        if (userOp.initCode.length != 0) {
-            _validateConstructor(userOp);
-            require(balanceOf(userOp.sender) >= tokenPrefund, "TokenPaymaster: no balance (pre-create)");
-        } else {
-
-            require(balanceOf(userOp.sender) >= tokenPrefund, "TokenPaymaster: no balance");
+    /// @notice Validates a paymaster user operation and calculates the required token amount for the transaction.
+    /// @param userOp The user operation data.
+    /// @param requiredPreFund The amount of tokens required for pre-funding.
+    /// @return context The context containing the token amount and user sender address (if applicable).
+    /// @return validationResult A uint256 value indicating the result of the validation (always 0 in this implementation).
+    function _validatePaymasterUserOp(UserOperation calldata userOp, bytes32, uint256 requiredPreFund)
+    internal
+    override
+    returns (bytes memory context, uint256 validationResult) {unchecked {
+            uint256 priceMarkup = tokenPaymasterConfig.priceMarkup;
+            uint256 paymasterAndDataLength = userOp.paymasterAndData.length - 20;
+            require(paymasterAndDataLength == 0 || paymasterAndDataLength == 32,
+                "TPM: invalid data length"
+            );
+            uint256 preChargeNative = requiredPreFund + (REFUND_POSTOP_COST * userOp.maxFeePerGas);
+        // note: as price is in ether-per-token and we want more tokens increasing it means dividing it by markup
+            uint256 cachedPriceWithMarkup = cachedPrice * PRICE_DENOMINATOR / priceMarkup;
+            if (paymasterAndDataLength == 32) {
+                uint256 clientSuppliedPrice = uint256(bytes32(userOp.paymasterAndData[20 : 52]));
+                if (clientSuppliedPrice < cachedPriceWithMarkup) {
+                    // note: smaller number means 'more ether per token'
+                    cachedPriceWithMarkup = clientSuppliedPrice;
+                }
+            }
+            uint256 tokenAmount = weiToToken(preChargeNative, cachedPriceWithMarkup, false);
+            SafeERC20.safeTransferFrom(token, userOp.sender, address(this), tokenAmount);
```

Comment:
minor, IIRC from our past debug experience, SafeERC20 uses banned [selfbalance opcode](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/dfef6a68ee18dbd2e1f5a099061a3b8a0e404485/contracts/utils/Address.sol#L134)

Reply:
Thanks, I did not know that! Created a small PR in the OpenZeppelin repo, hope this is all there is to this issue:
https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4271

I hope to avoid having to copy `SafeERC20.sol` library into the `account-abstraction` repo, but maybe we will have no choice.

Reply:
maybe we can use solady in that case https://github.com/Vectorized/solady/blob/f9a6cc652dc8b37387590588a78644abfa4be3ac/src/utils/SafeTransferLib.sol#L238